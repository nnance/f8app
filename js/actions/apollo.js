

const logError = require('logError');

export default function loadApolloQuery(type, query) {
  return dispatch => query.then((graphQLResult) => {
    const { errors, data } = graphQLResult;

    if (data) {
      if (Array.isArray(type)) {
        type.forEach(eachType => dispatch({ type: eachType, data }));
      } else {
        dispatch({ type, data });
      }
    }

    if (errors) {
      console.log('got some GraphQL execution errors', errors);
    }
  }).catch(logError);
}
