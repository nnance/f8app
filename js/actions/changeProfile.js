/* @flow */


import Parse from 'parse/react-native';
import _ from 'lodash';

import type { Action } from './types';

async function getUser() {
  const user = await Parse.User.currentAsync();
  if (!user) {
    throw new Error('not logged in');
  }
  return user;
}

async function _changePassword(newPassword) {
  const user = await getUser();
  user.set('password', newPassword);
  await user.save();
}

async function _changeEmail(newEmail) {
  const user = await getUser();
  user.set('username', newEmail);
  user.set('email', newEmail);
  await user.save();
}

function changedEmail(email): Action {
  return {
    type: 'CHANGED_EMAIL',
    payload: email,
  };
}

export const changeEmail = newEmail => dispatch => _changeEmail(newEmail).then(() => {
  dispatch(changedEmail(newEmail));
});

function changedPassword() {
  return {
    type: 'CHANGED_PASSWORD',
  };
}

export const changePassword = newPassword => dispatch => _changePassword(newPassword).then(() => {
  dispatch(changedPassword());
});

async function _changeProfile(data) {
  const user = await getUser();
  if (data.profilePicture) {
    // const base64 = `data:image/jpeg;base64,${data.profilePicture.data}`
    const base64 = data.profilePicture.data;
    const file = new Parse.File(data.profilePicture.fileName, { base64 });
    await file.save();
    user.set('profilePicture', file);
  }
  if (data.profileCover) {
    const base64 = data.profileCover.data;
    const file = new Parse.File(data.profileCover.fileName, { base64 });
    await file.save();
    user.set('profileCover', file);
  }
  const setData = _.omit(data, 'profileCover', 'profilePicture');
  Object.keys(setData).forEach((key) => {
    user.set(key, setData[key]);
  });
  // await new Promise(resolve => {
  //   setTimeout(() => resolve(), 2000);
  // })
  await user.save();
  const refetchUser = await user.fetch();
  return refetchUser;
}

export function changedPublicProfile(data) {
  return {
    type: 'CHANGED_PUBLIC_PROFILE',
    payload: data,
  };
}

export const changePublicProfile = (name, birthDayDate, sex, profilePicture, profileCover) =>
  dispatch =>
    _changeProfile({
      name,
      birthDayDate,
      sex,
      profilePicture,
      profileCover,
    }).then(
      (user) => {
        dispatch(changedPublicProfile({
          name: user.get('name'),
          sex: user.get('sex'),
          birthDayDate: user.get('birthDayDate'),
          profilePicture: user.get('profilePicture') ? user.get('profilePicture').url() : null,
          profileCover: user.get('profileCover') ? user.get('profileCover').url() : null,
        }));
      },
    );
