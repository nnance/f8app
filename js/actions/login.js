/**
 * Copyright 2016 Facebook, Inc.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to
 * use, copy, modify, and distribute this software in source code or binary
 * form for use in connection with the web services and APIs provided by
 * Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use
 * of this software is subject to the Facebook Developer Principles and
 * Policies [http://developers.facebook.com/policy/]. This copyright notice
 * shall be included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE
 *
 * @flow
 */

/* eslint no-unused-vars: warn */

import type { Action, ThunkAction } from './types';

const Parse = require('parse/react-native');
const FacebookSDK = require('FacebookSDK');
const ActionSheetIOS = require('ActionSheetIOS');
const { Platform } = require('react-native');
const Alert = require('Alert');
const { restoreSchedule, loadFriendsSchedules } = require('./schedule');
const { updateInstallation } = require('./installation');
const { loadSurveys } = require('./surveys');

async function ParseFacebookLogin(scope): Promise {
  return new Promise((resolve, reject) => {
    Parse.FacebookUtils.logIn(scope, {
      success: resolve,
      error: (user, error) => reject(error),
    });
  });
}

async function queryFacebookAPI(path, ...args): Promise {
  return new Promise((resolve, reject) => {
    FacebookSDK.api(path, ...args, (response) => {
      if (response && !response.error) {
        resolve(response);
      } else {
        reject(response && response.error);
      }
    });
  });
}

async function _logInWithFacebook(source: ?string): Promise<Array<Action>> {
  await ParseFacebookLogin('public_profile,email,user_friends');
  const profile = await queryFacebookAPI('/me', { fields: 'name,email' });

  const user = await Parse.User.currentAsync();
  user.set('facebook_id', profile.id);
  user.set('name', profile.name);
  user.set('email', profile.email);
  await user.save();
  await updateInstallation({ user });

  const action = {
    type: 'LOGGED_IN',
    source,
    data: {
      id: profile.id,
      name: profile.name,
      sex: user.get('sex'),
      sharedSchedule: user.get('sharedSchedule'),
      birthDayDate: user.get('birthDayDate'),
      profilePicture: user.get('profilePicture') ? user.get('profilePicture').url() : null,
      profileCover: user.get('profileCover') ? user.get('profileCover').url() : null,
      facebookLinked: true,
    },
  };

  return Promise.all([
    Promise.resolve(action),
    restoreSchedule(),
  ]);
}

function signedUp(): Action {
  return {
    type: 'SIGNED_UP',
  };
}

const signUp = (email: string, password: string) => (dispatch) => {
  const user = new Parse.User();
  user.set('username', email);
  user.set('name', email);
  user.set('password', password);
  user.set('email', email);
  user.set('birthDayDate', null);
  return user.save().then(() => {
    dispatch(signedUp());
  });
};

async function _linkFacebook(user) {
  await new Promise((resolve, reject) => {
    Parse.FacebookUtils.link(user, null, {
      success: () => resolve(),
      error: (u, error) => reject(error.error),
    });
  });
  await user.save();
  const profile = await queryFacebookAPI('/me', { fields: 'name,email' });
  user.set('facebook_id', profile.id);
  user.set('name', profile.name);
  await user.save();
  await updateInstallation({ user });
  return user;
}

async function _unlinkFacebook(user) {
  await new Promise((resolve, reject) => {
    Parse.FacebookUtils.unlink(user, {
      success: () => resolve(),
      error: error => reject(error),
    });
  });
  await user._unlinkFrom('facebook');
  user.set('facebook_id', null);
  await user.save();
  await updateInstallation({ user });
  return user;
}

const facebookLinked = (id, name) => ({
  type: 'FACEBOOK_LINKED',
  data: {
    id,
    name,
  },
});

const facebookUnlinked = () => ({
  type: 'FACEBOOK_UNLINKED',
});

const linkFacebook = () => async (dispatch) => {
  const user = await Parse.User.currentAsync();
  return _linkFacebook(user).then((_user) => {
    dispatch(facebookLinked(_user.get('facebook_id'), _user.get('name')));
  });
};

const unlinkFacebook = () => async (dispatch) => {
  const user = await Parse.User.currentAsync();
  return _unlinkFacebook(user).then(() => {
    dispatch(facebookUnlinked());
  });
};

const logIn = (email: string, password: string) => dispatch =>
  Parse.User.logIn(email, password).then(
    async () => {
      const user = await Parse.User.currentAsync();
      await updateInstallation({ user });
      const action = {
        type: 'LOGGED_IN',
        data: {
          id: null,
          name: user.get('name'),
          sex: user.get('sex'),
          birthDayDate: user.get('birthDayDate'),
          sharedSchedule: user.get('sharedSchedule'),
          profilePicture: user.get('profilePicture') ? user.get('profilePicture').url() : null,
          profileCover: user.get('profileCover') ? user.get('profileCover').url() : null,
          facebookLinked: user.get('id') !== null && user.get('id') !== undefined,
        },
      };
      dispatch(action);
      dispatch(restoreSchedule());
    },
  );

function logInWithFacebook(source: ?string): ThunkAction {
  return (dispatch) => {
    const login = _logInWithFacebook(source);

    // Loading friends schedules shouldn't block the login process
    login.then(
      (result) => {
        dispatch(result);
        dispatch(loadFriendsSchedules());
        dispatch(loadSurveys());
      },
    );
    return login;
  };
}

function skipLogin(): Action {
  return {
    type: 'SKIPPED_LOGIN',
  };
}

const forgotPassword: ThunkAction = email => dispatch => Parse.User.requestPasswordReset(email);

function logOut(): ThunkAction {
  return (dispatch) => {
    Parse.User.logOut();
    FacebookSDK.logout();
    updateInstallation({ user: null, channels: [] });

    // TODO: Make sure reducers clear their state
    return dispatch({
      type: 'LOGGED_OUT',
    });
  };
}

function logOutWithPrompt(): ThunkAction {
  return (dispatch, getState) => {
    const name = getState().user.name || 'there';

    if (Platform.OS === 'ios') {
      ActionSheetIOS.showActionSheetWithOptions(
        {
          title: `Hi, ${name}`,
          options: ['Log out', 'Cancel'],
          destructiveButtonIndex: 0,
          cancelButtonIndex: 1,
        },
        (buttonIndex) => {
          if (buttonIndex === 0) {
            dispatch(logOut());
          }
        },
      );
    } else {
      Alert.alert(
        `Hi, ${name}`,
        'Log out from F8?',
        [
          { text: 'Cancel' },
          { text: 'Log out', onPress: () => dispatch(logOut()) },
        ],
      );
    }
  };
}

module.exports = {
  logInWithFacebook,
  skipLogin,
  forgotPassword,
  signUp,
  logIn,
  logOut,
  logOutWithPrompt,
  linkFacebook,
  unlinkFacebook,
  facebookUnlinked,
  facebookLinked,
};
