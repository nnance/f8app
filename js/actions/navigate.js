export function navigate(route) {
  return {
    type: 'NAV_TO',
    payload: route,
  };
}

export function navigateBack() {
  return {
    type: 'NAV_BACK',
  };
}
