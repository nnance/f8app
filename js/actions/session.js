

import gql from 'graphql-tag';
import apollo from '../store/apollo';

import type { ThunkAction } from './types';

import loadApolloQuery from './apollo';


function loadSessions() : ThunkAction {
  const query = apollo.query({
    query: gql`
      query sessions {
        schedule {
          id
          title
          slug
          day
          startTime
          endTime
          location {
            id
            name
            x1url
            x2url
            x3url
          }
          description
          speakers {
            id
            name
            title
            picture
          }
          isAdded
        }
      }
    `,
    forceFetch: false,
  });

  return loadApolloQuery('LOADED_SESSIONS', query);
}

module.exports = {
  loadSessions,
};
