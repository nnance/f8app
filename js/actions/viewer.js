

import gql from 'graphql-tag';
import apollo from '../store/apollo';

import type { ThunkAction } from './types';

import loadApolloQuery from './apollo';


function loadViewer() : ThunkAction {
  const query = apollo.query({
    query: gql`
      query viewer {
        viewer {
          faqs {
            id
            question
            answer
          }
          pages {
            id
            title
            url
            logo
          }
          maps{
            id
            name
            x1url
            x2url
            x3url
          }
        }
      }
    `,
    forceFetch: false,
  });

  const actions = ['LOADED_PAGES', 'LOADED_MAPS', 'LOADED_FAQS'];
  return loadApolloQuery(actions, query);
}

module.exports = {
  loadViewer,
};
