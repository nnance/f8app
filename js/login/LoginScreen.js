

import React from 'react';
import { BackAndroid, Text, TouchableOpacity, Navigator, Image, View } from 'react-native';
import { connect } from 'react-redux';

import * as actions from '../actions';
import { withTracking } from '../common/navigateTracking';

import IndexScreen from './IndexScreen';
import EmailLoginScreen from './EmailLoginScreen';
import SignupScreen from './SignupScreen';
import ForgotPasswordScreen from './ForgotPasswordScreen';
import SuccessScreen from './SuccessScreen';
import styles from './styles';

const titles = {
  index: '',
  'email-login': 'ลงชื่อเข้าใช้ผ่านอีเมล',
  signup: 'สร้างบัญชีใหม่',
  forgotPassword: 'ขอรหัสผ่านใหม่',
  success: '',
};

class LoginScreen extends React.Component {
  constructor(args) {
    super(...args);
    this.goToLogin = this.goToLogin.bind(this);
    this.pushPage = this.pushPage.bind(this);
    this.goBack = this.goBack.bind(this);
    this.renderScene = this.renderScene.bind(this);
    this.logInWithFacebook = this.logInWithFacebook.bind(this);
    this.renderLeftButton = this.renderLeftButton.bind(this);
    this.renderTitle = this.renderTitle.bind(this);
    this.renderRightButton = this.renderRightButton.bind(this);
    this.renderScene = this.renderScene.bind(this);
  }

  componentDidMount() {
    BackAndroid.addEventListener('hardwareBackPress', this.goBack);
    if (this.props.addBackButtonListener) {
      this.props.addBackButtonListener(this.alwaysFalse);
    }
    this.props.navigate('login');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.isLoggedIn && this.props.onExit) {
      this.props.onExit();
    }
  }

  componentWillUnmount() {
    BackAndroid.removeEventListener('hardwareBackPress', this.goBack);
    if (this.props.removeBackButtonListener) {
      this.props.removeBackButtonListener(this.alwaysFalse);
    }
  }

  alwaysFalse() {
    return false;
  }

  goToLogin() {
    const navigator = this.navigator;
    const currentRoutes = navigator.getCurrentRoutes();
    let N = navigator.getCurrentRoutes().length;
    while (N) {
      if (currentRoutes[N - 1].page === 'email-login') {
        break;
      }
      N -= 1;
    }
    if (N <= 0) {
      throw Error('email-login route not found');
    }
    navigator.popN(navigator.getCurrentRoutes().length - N);
    this.props.navigate('login');
  }

  pushPage(page, payload) {
    this.navigator.push({ page, payload });
    this.props.navigate(`login/${page}`);
  }

  goBack() {
    this.props.navigateBack();
    if (this.navigator.getCurrentRoutes().length > 1) {
      this.navigator.pop();
      return true;
    } else if (this.props.onExit) {
      this.props.onExit();
      return true;
    }
    return false;
  }

  async logInWithFacebook() {
    const { dispatch, onLoggedIn } = this.props;

    try {
      await dispatch(actions.logInWithFacebook());
    } catch (e) {
      const message = e.message || e;
      if (message !== 'Timed out' && message !== 'Canceled by user') {
        alert(message);
        console.warn(e);
      }
      return;
    }

    if (onLoggedIn) {
      onLoggedIn();
    }
  }

  renderTitle(route) {
    let title = '';
    title = titles[route.page];
    return (<View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}><Text style={styles.titleNavBar}>{title}</Text></View>);
  }

  renderLeftButton(route) {
    if (route.page === 'index') {
      return null;
    }
    return (
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
        <TouchableOpacity onPress={this.goBack}>
          <Image style={styles.backButton} source={require('../assets/common/white-back-button.png')} />
        </TouchableOpacity>
      </View>
    );
  }

  renderRightButton(route) {
    if (route.page !== 'index') {
      return null;
    }
    return (
      <TouchableOpacity
        accessibilityLabel="Skip login"
        accessibilityTraits="button"
        style={styles.skip}
        onPress={() => this.props.skipLogin()}
      >
        <Image
          source={require('../assets/login/x.png')}
        />
      </TouchableOpacity>
    );
  }

  renderScene(route) {
    if (route.page === 'index') {
      return <IndexScreen {...this.props} pushPage={this.pushPage} />;
    }
    if (route.page === 'email-login') {
      return (
        <EmailLoginScreen
          error={this.props.error}
          logIn={this.props.logIn}
          pushPage={this.pushPage}
        />
      );
    }
    if (route.page === 'signup') {
      return (<SignupScreen
        error={this.props.error}
        signUp={this.props.signUp}
        logIn={this.props.logIn}
        goBack={this.goBack}
        logInWithFacebook={this.logInWithFacebook}
      />);
    }
    if (route.page === 'forgotPassword') {
      return (<ForgotPasswordScreen
        error={this.props.error}
        forgotPassword={this.props.forgotPassword}
        goBack={this.goBack}
        onReqed={() => this.pushPage('success', { successText: 'mail has been sent' })}
      />);
    }
    if (route.page === 'success') {
      return <SuccessScreen successText={route.payload.successText} goToLogin={this.goToLogin} />;
    }
    return <Text>Page not found</Text>;
  }

  render() {
    return (
      <Navigator
        ref={
          (node) => {
            this.navigator = node;
          }
        }
        initialRoute={{ page: this.props.withEmail ? 'email-login' : 'index' }}
        renderScene={this.renderScene}
        navigationBar={
          <Navigator.NavigationBar
            navigationStyles={Navigator.NavigationBar.StylesIOS}
            routeMapper={{
              LeftButton: this.renderLeftButton,
              RightButton: this.renderRightButton,
              Title: this.renderTitle,
            }}
          />
        }
      />
    );
  }
}

const select = state => ({
  isLoggedIn: state.user.isLoggedIn,
  isSignedUp: state.user.isSignedUp,
  error: state.user.loginError,
});

const actionsMaping = {
  skipLogin: actions.skipLogin,
  logIn: actions.logIn,
  signUp: actions.signUp,
  forgotPassword: actions.forgotPassword,
};

export default withTracking(connect()(connect(select, actionsMaping)(LoginScreen)));
export {
  LoginScreen as Component,
};
