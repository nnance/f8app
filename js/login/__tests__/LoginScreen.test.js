jest.mock('BackAndroid');

/* eslint import/first: off */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import {
  View,
  BackAndroid,
} from 'react-native';

jest.mock('../../actions');

import { Component as LoginScreenComponent } from '../LoginScreen';
import * as actions from '../../actions';

describe('LoginScreen', () => {
  it('render', () => {
    const tree = shallow(<LoginScreenComponent />);
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('renderTitle', () => {
    const dump = shallow(<LoginScreenComponent />);
    const tree = renderer.create(dump.instance().renderTitle({ page: 'signup' }));
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renderLeftButton', () => {
    const dump = shallow(<LoginScreenComponent />);
    let tree = renderer.create(dump.instance().renderLeftButton({ page: 'signup' }));
    expect(tree.toJSON()).toMatchSnapshot();
    tree = renderer.create(<View>{dump.instance().renderLeftButton({ page: 'index' })}</View>);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renderRightButton', () => {
    const dump = shallow(<LoginScreenComponent />);
    let tree = renderer.create(<View>{dump.instance().renderRightButton({ page: 'signup' })}</View>);
    expect(tree.toJSON()).toMatchSnapshot();
    tree = renderer.create(<View>{dump.instance().renderRightButton({ page: 'index' })}</View>);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renderScene', () => {
    const dump = shallow(<LoginScreenComponent />).instance();
    const routes = [
      { page: 'index' },
      { page: 'email-login' },
      { page: 'signup' },
      { page: 'forgotPassword' },
      { page: 'success',
        payload: 'test' },
      {},
    ];
    routes.forEach((route) => {
      const tree = shallow(<View>{dump.renderScene(route)}</View>);
      expect(toJSON(tree)).toMatchSnapshot();
    });
  });

  it('listen to BackAndroid for handleBackButton', () => {
    const wrapper = shallow(<LoginScreenComponent navigate={jest.fn()} />);
    wrapper.instance().componentDidMount();
    expect(BackAndroid.addEventListener).toBeCalledWith('hardwareBackPress', wrapper.instance().goBack);
    wrapper.instance().componentWillUnmount();
    expect(BackAndroid.removeEventListener).toBeCalledWith('hardwareBackPress', wrapper.instance().goBack);
  });

  it('addBackButtonListener alwayFalse', () => {
    const addSpy = jest.fn();
    const removeSpy = jest.fn();
    const wrapper = shallow(
      <LoginScreenComponent
        addBackButtonListener={addSpy}
        removeBackButtonListener={removeSpy}
        navigate={jest.fn()}
      />,
    );
    wrapper.instance().componentDidMount();
    expect(addSpy).toBeCalledWith(wrapper.instance().alwaysFalse);
    wrapper.instance().componentWillUnmount();
    expect(removeSpy).toBeCalledWith(wrapper.instance().alwaysFalse);
  });

  it('auto call onExit if nextProps is loggedIn', () => {
    const spy = jest.fn();
    const wrapper = shallow(<LoginScreenComponent onExit={spy} />);
    wrapper.instance().componentWillReceiveProps({ isLoggedIn: true });
    expect(spy).toBeCalled();
  });

  it('goBack will pop navigator', () => {
    const wrapper = shallow(<LoginScreenComponent navigateBack={jest.fn()} />);
    const spy = jest.fn();
    const inst = wrapper.instance();
    inst.navigator = {
      getCurrentRoutes: jest.fn(() => [1, 1, 1]),
      pop: spy,
    };
    inst.goBack();
    expect(spy).toBeCalled();
  });

  it('goBack will call onExit if on LastPage', () => {
    const spy = jest.fn();
    const wrapper = shallow(<LoginScreenComponent onExit={spy} navigateBack={jest.fn()} />);
    const inst = wrapper.instance();
    inst.navigator = {
      getCurrentRoutes: jest.fn(() => [1]),
      pop: jest.fn(),
    };
    inst.goBack();
    expect(spy).toBeCalled();
  });

  it('goToLogin pop untill login-page', () => {
    const wrapper = shallow(<LoginScreenComponent navigate={jest.fn()} navigateBack={jest.fn()} />);
    const inst = wrapper.instance();
    const pages = [{ page: 'email-login' }, { page: 'fake-1' }, { page: 'fake-2' }];
    const spy = jest.fn(pages.pop);
    pages.pop = spy;
    inst.navigator = {
      getCurrentRoutes: jest.fn(() => pages),
      popN: (n) => {
        let N = n;
        while (N) {
          N -= 1;
          pages.pop();
        }
      },
    };
    inst.goToLogin();
    expect(pages.length).toBe(1);
    expect(pages[0]).toEqual({ page: 'email-login' });
    expect(spy).toHaveBeenCalledTimes(2);
  });

  describe('logInWithFacebook', () => {
    let onLoggedIn;
    let alertSpy;
    let preAlert;
    let preWarn;

    beforeAll(() => {
      preAlert = global.alert;
      preWarn = global.console.warn;
      alertSpy = jest.fn();
      onLoggedIn = jest.fn();

      global.console.warn = jest.fn();
      global.alert = alertSpy;
    });

    afterAll(() => {
      global.console.warn = preWarn;
      global.alert = preAlert;
    });

    afterEach(() => {
      alertSpy.mockClear();
      onLoggedIn.mockClear();
      actions.logInWithFacebook.mockClear();
    });

    it('call onLoggedIn if success', async () => {
      actions.logInWithFacebook.mockImplementationOnce(
        () => Promise.resolve(),
      );
      const wrapper = shallow(
        <LoginScreenComponent
          onLoggedIn={onLoggedIn}
          dispatch={jest.fn(a => a)}
        />,
      );
      const inst = wrapper.instance();
      await inst.logInWithFacebook();
      expect(actions.logInWithFacebook).toBeCalled();
      expect(onLoggedIn).toBeCalled();
    });

    it('call alert if fail', async () => {
      actions.logInWithFacebook.mockImplementationOnce(
        () => Promise.reject(new Error('something wrong')),
      );
      const wrapper = shallow(
        <LoginScreenComponent
          onLoggedIn={onLoggedIn}
          dispatch={jest.fn(a => a)}
        />,
      );
      const inst = wrapper.instance();
      await inst.logInWithFacebook();
      expect(onLoggedIn).not.toBeCalled();
      expect(alertSpy).toBeCalled();
    });
  });
});
