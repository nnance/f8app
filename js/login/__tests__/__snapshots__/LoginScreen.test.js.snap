exports[`LoginScreen render 1`] = `
<Navigator
  configureScene={[Function]}
  initialRoute={
    Object {
      "page": "index",
    }
  }
  navigationBar={
    <NavigatorNavigationBar
      navigationStyles={
        Object {
          "General": Object {
            "NavBarHeight": 44,
            "StatusBarHeight": 20,
            "TotalNavHeight": 64,
          },
          "Interpolators": Object {
            "CenterToLeft": Object {
              "LeftButton": [Function],
              "RightButton": [Function],
              "Title": [Function],
            },
            "RightToCenter": Object {
              "LeftButton": [Function],
              "RightButton": [Function],
              "Title": [Function],
            },
            "RightToLeft": Object {
              "LeftButton": [Function],
              "RightButton": [Function],
              "Title": [Function],
            },
          },
          "Stages": Object {
            "Center": Object {
              "LeftButton": Object {
                "backgroundColor": "transparent",
                "height": 44,
                "left": 0,
                "opacity": 1,
                "overflow": "hidden",
                "position": "absolute",
                "top": 20,
              },
              "RightButton": Object {
                "alignItems": "flex-end",
                "backgroundColor": "transparent",
                "height": 44,
                "opacity": 1,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
              "Title": Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "height": 44,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
            },
            "Left": Object {
              "LeftButton": Object {
                "backgroundColor": "transparent",
                "height": 44,
                "left": 0,
                "opacity": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 20,
              },
              "RightButton": Object {
                "alignItems": "flex-end",
                "backgroundColor": "transparent",
                "height": 44,
                "opacity": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
              "Title": Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "height": 44,
                "left": -375,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
            },
            "Right": Object {
              "LeftButton": Object {
                "backgroundColor": "transparent",
                "height": 44,
                "left": 0,
                "opacity": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 20,
              },
              "RightButton": Object {
                "alignItems": "flex-end",
                "backgroundColor": "transparent",
                "height": 44,
                "opacity": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
              "Title": Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "height": 44,
                "left": 375,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 20,
              },
            },
          },
        }
      }
      routeMapper={
        Object {
          "LeftButton": [Function],
          "RightButton": [Function],
          "Title": [Function],
        }
      } />
  }
  renderScene={[Function]}
  sceneStyle={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
        Object {
          "translateY": 0,
        },
        Object {
          "scaleX": 1,
        },
        Object {
          "scaleY": 1,
        },
        Object {
          "rotate": "0deg",
        },
        Object {
          "skewX": "0deg",
        },
        Object {
          "skewY": "0deg",
        },
      ],
    }
  } />
`;

exports[`LoginScreen renderLeftButton 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}>
    <Image
      source={1}
      style={
        Object {
          "height": 20,
          "paddingHorizontal": 15,
          "resizeMode": "contain",
          "width": 20,
        }
      } />
  </View>
</View>
`;

exports[`LoginScreen renderLeftButton 2`] = `<View />`;

exports[`LoginScreen renderRightButton 1`] = `<View />`;

exports[`LoginScreen renderRightButton 2`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel="Skip login"
    accessibilityTraits="button"
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "padding": 15,
      }
    }
    testID={undefined}>
    <Image
      source={1} />
  </View>
</View>
`;

exports[`LoginScreen renderScene 1`] = `
<View>
  <Connect(IndexScreen)
    pushPage={[Function]} />
</View>
`;

exports[`LoginScreen renderScene 2`] = `
<View>
  <EmailLoginScreen
    pushPage={[Function]} />
</View>
`;

exports[`LoginScreen renderScene 3`] = `
<View>
  <SignUpScreen
    goBack={[Function]}
    logInWithFacebook={[Function]} />
</View>
`;

exports[`LoginScreen renderScene 4`] = `
<View>
  <ForgotPasswordScreen
    goBack={[Function]}
    onReqed={[Function]} />
</View>
`;

exports[`LoginScreen renderScene 5`] = `
<View>
  <Component
    goToLogin={[Function]} />
</View>
`;

exports[`LoginScreen renderScene 6`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Page not found
  </Text>
</View>
`;

exports[`LoginScreen renderTitle 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "white",
        "fontWeight": "bold",
        "padding": 10,
      }
    }>
    สร้างบัญชีใหม่
  </Text>
</View>
`;
